core/schema/env.graphqls

```diff
@@ -0,0 +1,252 @@
+  """
+  from(
+    name: String!
+    sourceDirectory: DirectoryID!
+    sdk: String!
+    sourceDirectorySubpath: String
+    dependencies: [EnvironmentID!]
+  ): Environment!
+
+  """
+  The environment initialized with the sourceDirectory and environment configuration file path.
+  If configPath is not set, it defaults to the root of the sourceDirectory.
+  """
+  fromConfig(sourceDirectory: DirectoryID!, configPath: String): Environment!
+
+  """
+  The directory that the environment code will execute in as its current working directory.
+  If not set explicitly, it will default to the root of the sourceDirectory.
+  """
+  workdir: Directory!
+
+  "This environment with the given workdir"
+  withWorkdir(id: DirectoryID!): Environment!
+
+  "The list of checks in this environment"
+  checks: [Check!]
+
+  "The check in this environment with the given name, if any"
+  check(name: String!): Check
+}
+
+"An entrypoint for tests, lints or anything that can pass/fail."
+type Check {
+  "A unique identifier for this check."
+  id: CheckID!
+
+  "The name of the check."
+  name: String!
+
+  "This check with the given name"
+  withName(name: String!): Check!
+
+  "Documentation for this check."
+  description: String
+
+  "This check with the given description"
+  withDescription(description: String!): Check!
+
+  "The subchecks of this check."
+  subchecks: [Check!]
+
+  "This check with the given subcheck"
+  withSubcheck(id: CheckID!): Check!
+
+  """
+  This check with the given container used to determine the check's result.
+  If set, the container will be executed and the check success will be true
+  if the container exits with a zero exit code, false otherwise.
+  """
+  withContainer(id: ContainerID!): Check!
+
+  "The result of evaluating this check."
+  result: CheckResult!
+}
+
+type CheckResult {
+  "A unique identifier for this check result."
+  id: CheckResultID!
+
+  "Whether the check was successful."
+  success: Boolean!
+
+  "Any output obtained from evaluating the check's success."
+  output: String!
+}
+
+extend type Environment {
+  """
+  Internal only.
+
+  This environment with the given check. It is an error to call this outside
+  of environment initialization code.
+  """
+  withCheck(id: CheckID!, returnType: CheckEntrypointReturnType): Environment!
+
+  """
+  Internal only.
+
+  The input for the current entrypoint invocation.
+  """
+  entrypointInput: EntrypointInput!
+
+  """
+  Internal only.
+
+  Return the given value as the result of the current entrypoint invocation.
+  The value is expected to be the json-serialized representation of the return.
+  """
+  returnEntrypointValue(value: String!): Boolean!
+}
+
+"""
+Internal-only.
+
+The input provided to an SDK entrypoint invocation.
+"""
+type EntrypointInput {
+  """
+  Internal-only.
+
+  The name of the entrypoint to invoke. If not set, then the sdk
+  is expected to return the environment definition.
```

Comment:
How about adding an `Entrypoint.sync` field to load the `with_check` calls, etc, instead of via `returnEntrypointValue`?

Pseudo code (serve):
```python
env = dagger.current_environment()
input_ = env.entrypoint_input()
name = await input_.name()

if not name:
    await env.sync()
    return

...
```

Reply:
I actually started there; there were a few tricky parts because when we `Sync` this object in particular, the ID actually changes, so we hit the case you described [here](https://github.com/sipsma/dagger/blob/21e8e442736b0bbd301a8aca8e4b81e36fc309ad/codegen/generator/functions.go#L134-L142).

Those would be solvable with extra effort, but I ultimately went with the `returnEntrypointValue` approach anyways since not only avoids that whole problem but also ends up minimizing SDK responsibilities, moves as much to the engine as possible and gives us long term flexibility to modify how this works without updating every SDK:
1. It's important for the implementation that the result we return is actually written to filesystem of the entrypoint exec (so that the result is cached as part of that exec).
   * Previously, that meant every SDK needed to know where to write that file. Now, they are agnostic to it; they just need to call that API.
2. Centralizing what we write there entirely server-side is nice since we can very easily update it as needed, without any considerations around whether we'd impact the (non-internal) API of `.Sync`. It's highly plausible we'll want to make updates to the format of what we write or the details in how we process it since this is all intertwined with how the caching works.

Reply:
Yeah I understand but it can be any other name like `Environment.load` that's marked internal. Otherwise this looks a bit odd:

```python
await env.environment_result_value(str(await env.id()))
```

You have to serialize/marshal in the SDK, etc, would be simpler to just:

```python
await env.load()
```

I was confused if I can use the same environment instance because go sends the result value from a clean `dag.CurrentEnvironment()`. Is this required?